README.md

![Swift 4.0+](https://img.shields.io/badge/Swift-4.0+-orange.svg)
![Swift 3.2](https://img.shields.io/badge/Swift-3.2-orange.svg)
![Build](https://img.shields.io/badge/build-passing-green.svg)
![License](https://img.shields.io/cocoapods/l/SwiftyXMLParser.svg?style=flat)
![Platform](https://img.shields.io/badge/Platform-iOS10%2B-green.svg)

#ASXMLParser

One-Stop XML Parsing solution in swift

# What's this?
This is a XML parser inspired by [XML-Parser by Troy Brant](http://troybrant.net/blog/2010/09/simple-xml-to-nsdictionary-converter/).

[NSXMLParser](https://developer.apple.com/library/mac/documentation/Cocoa/Reference/Foundation/Classes/NSXMLParser_Class/) is inbuild XML parsing solution which is available in iOS and part of Foundation framework & is a kind of "SAX" parser. Which is a great framework but its not easy to use. 

# Feature
-  Parse XML Doument as NSDictionary.
-  Create XML from Dictionary.

# Requirement
+ iOS 8.0+
+ Swift 4.0+ (or Swift 3.2)

# Installation

#### 2. install

### CocoaPods
#### 1. create Podfile
```ruby:Podfile
platform :ios, '8.0'
use_frameworks!

pod "ASXMLParser"
```

#### 2. install
```
> pod install
```

# Example

```swift
    let xmlString = """
        <?xml version="1.0" encoding="UTF-8"?>
        <Books>
        <Book id="1">
        <title>Circumference</title>
        <author>Nicholas Nicastro</author>
        <summary>Eratosthenes and the Ancient Quest to Measure the Globe.</summary>
        </Book>
        <Book id="2">
        <title>Copernicus Secret</title>
        <author>Jack Repcheck</author>
        <summary>How the scientific revolution began</summary>
        </Book>
        <Book id="3">
        <title>Angels and Demons</title>
        <author>Dan Brown</author>
        <summary>Robert Langdon is summoned to a Swiss research facility to analyze a cryptic symbol seared into the chest of a murdered physicist.</summary>
        </Book>
        <Book id="4">
        <title>Keep the Aspidistra Flying</title>
        <author>George Orwell</author>
        <summary>A poignant and ultimately hopeful look at class and society, Keep the Aspidistra Flying pays tribute to the stubborn virtues of ordinary people who keep the aspidistra flying.</summary>
        </Book>
        </Books>
        """
        
        if let dictionary = XML().dictionaryForXMLString(string: xmlString) {
            print(dictionary)
        }
```

Output be like

```
 {
    Books = {
        Book = (
                		{   
                author = "Nicholas Nicastro";
                id = 1;
                summary = "Eratosthenes and the Ancient Quest to Measure the Globe.";
                title = Circumference;
            },
                        {
                author = "Jack Repcheck";
                id = 2;
                summary = "How the scientific revolution began";
                title = "Copernicus Secret";
            },
                        {
                author = "Dan Brown";
                id = 3;
                summary = "Robert Langdon is summoned to a Swiss research facility to analyze a cryptic symbol seared into the chest of a murdered physicist.";
                title = "Angels and Demons";
            },
                        {
                author = "George Orwell";
                id = 4;
                summary = "A poignant and ultimately hopeful look at class and society, Keep the Aspidistra Flying pays tribute to the stubborn virtues of ordinary people who keep the aspidistra flying.";
                title = "Keep the Aspidistra Flying";
            }
        );
    };
}
```

instead of using XML strings directly you can pass data object also, 

```
if let path = Bundle.main.path(forResource: "Foods", ofType: "xml"){
     let url = URL(fileURLWithPath: path)
     let xmlString = try! String(contentsOf: url, encoding: String.Encoding.utf8)

     if let dictionary = XML().dictionaryForXMLString(string: xmlString) {
         print(dictionary)
     }
}

```

Now, lets see how we can create XML from Dictionary,

```
let education = [["course":"graduation",
                          "institutation":"IITD",
                          "from":"2007",
                          "to":"2011"],
                         ["course":"10+2",
                          "institutation":"DPS",
                          "from":"2005",
                          "to":"2007"],
                         ["course":"10",
                          "institutation":"DPS",
                          "from":"2000",
                          "to":"2005"]]
        let dic : [String : Any] = ["fisrtname":"John",
                                    "lastname":"Doe",
                                    "email":"john@doe.com",
                                    "age":"36",
                                    "gender":"male",
                                    "education":education]
        let xml = ASXMLCreator.init(input: dic)
        print("XML: \(xml.output ?? "")")
```

Output will be, 

```
XML: <?xml version="1.0" encoding="UTF-8" ?>
		<root>
		<fisrtname>John</fisrtname>
		<age>36<age>
		<email>john@doe.com</email>
		<lastname>Doe</lastname>
		<gender>male</gender>
		<education>
			<from>2007</from>
			<institutation>IITD</institutation>
			<course>graduation</course>
			<to>2011</to>
			</education>
		<education>
			<from>2005</from>
			<institutation>DPS</institutation>
			<course>10+2</course>
			<to>2007</to>
			</education>
		<education>
			<from>2000</from>
			<institutation>DPS</institutation>
			<course>10</course>
			<to>2005</to>
		</education>
		</root>
```
p.s: this is formatted output, orogially it will be single line output.  

# License

This software is released under the MIT License, see LICENSE.


CHANGES.md
# Changelog

### 1.0.0 (26/06/2018)

* First release




XMLParser/ASXMLParser.swift
//
//  Parser.swift
//  XMLParser
//
//  Created by Amitabha Saha on 25/06/18.
//  Copyright © 2018 ASTech. All rights reserved.
//

// Playground - noun: a place where people can play

import Foundation

public enum XMLReaderOptions: UInt {
    case ZeroValue = 0
    case ProcessNamespaces = 1
    case ReportNamespacePrefixes = 2
    case ResolveExternalEntities = 4
}

let kXMLReaderTextNodeKey: NSString = "text"
let kXMLReaderAttributePrefix: NSString = "@"

public class ASXMLParser: NSObject, XMLParserDelegate {
    
    var dictionaryStack: NSMutableArray?
    var textInProgress = String()
    
    
    open func dictionaryForXMLData(data: Data) -> NSDictionary? {
        
        let rootDictionary: NSDictionary? = self.objectWithData(data: data, options: XMLReaderOptions.ZeroValue)
        return rootDictionary
        
    }
    
    open func dictionaryForXMLString(string: String) -> NSDictionary? {
        if let data: Data = string.data(using: .utf8) {
            return dictionaryForXMLData(data: data)
        } else {
            return nil
        }
    }
    
    open func dictionaryForXMLData(data: Data, options: XMLReaderOptions) -> NSDictionary? {
        
        let rootDictionary: NSDictionary? = self.objectWithData(data: data, options: options)
        return rootDictionary
    }
    
    open func dictionaryForXMLString(string: NSString, options: XMLReaderOptions) -> NSDictionary? {
        
        if let data: Data = string.data(using: String.Encoding.utf8.rawValue) {
            return dictionaryForXMLData(data: data, options: options)
        } else {
            return nil
        }
    }
    
    func objectWithData(data: Data, options: XMLReaderOptions) -> NSDictionary? {
        
        // Clear out any old Data
        dictionaryStack = NSMutableArray()
        textInProgress = String()
        
        // Initialize the stack with a fresh dictionary
        dictionaryStack?.add(NSMutableDictionary())
        let parser: XMLParser = XMLParser(data: data)
        
        parser.shouldProcessNamespaces = (Bool(0 != options.rawValue & XMLReaderOptions.ProcessNamespaces.rawValue))
        parser.shouldReportNamespacePrefixes = (Bool(0 != options.rawValue & XMLReaderOptions.ReportNamespacePrefixes.rawValue))
        parser.shouldResolveExternalEntities = (Bool(0 != options.rawValue & XMLReaderOptions.ResolveExternalEntities.rawValue))
        
        parser.delegate = self
        
        let success: Bool = parser.parse()
        if success {
            let resultDict: NSDictionary = dictionaryStack?.firstObject as! NSDictionary
            if let resultFiltered = filterResponse(resultDict as? NSMutableDictionary) {
                return resultFiltered
            } else {
                return nil
            }
        }
        return nil
    }
    
    public func parser(_ parser: XMLParser, didStartElement elementName: String, namespaceURI: String?, qualifiedName qName: String?, attributes attributeDict: [String : String] = [:]) {
        
        // Get the dictionary for the current level in the stack
        let parentDict = self.dictionaryStack?.lastObject as! NSMutableDictionary
        
        // Create the child dictionary for the new element, and initilaize it with the attributes
        let childDict = NSMutableDictionary()
        childDict.addEntries(from: attributeDict)
        
        // If there's already an item for this key, it means we need to create an array
        let existingValue = parentDict[elementName]
        
        if existingValue != nil {
            
            var array: NSMutableArray? = nil
            
            if let arrayValue =  existingValue as? NSMutableArray{
                
                // The array exists, so use it
                array = arrayValue
            } else {
                
                // Create an array if it doesn't exist
                array = NSMutableArray()
                array?.add(existingValue!)
                
                // Replace the child dictionary with an array of children dictionaries
                parentDict.setObject(array!, forKey: elementName as NSCopying)
            }
            
            // Add the new child dictionary to the array
            array?.add(childDict)
            
        } else {
            // Add the new child dictionary to the array
            parentDict.setObject(childDict, forKey: elementName as NSCopying)
        }
        
        // No existing value, so update the dictionary
        self.dictionaryStack?.add(childDict)
        
    }
    
    
    public func parser(_ parser: XMLParser, foundCharacters string: String) {
        textInProgress += string
    }
    
    public func parser(_ parser: XMLParser, didEndElement elementName: String, namespaceURI: String?, qualifiedName qName: String?) {
        
        // Update the parent dict with text info
        if let dictInProgress: NSMutableDictionary = self.dictionaryStack?.lastObject as? NSMutableDictionary{
            
            if self.textInProgress.count > 0 {
                
                let trimmedString = self.textInProgress.trimmingCharacters(in: .whitespacesAndNewlines)
                dictInProgress.setObject(trimmedString, forKey: elementName as NSCopying)
                self.textInProgress = String()
            }
            
        }
        
        // Pop the current dict
        self.dictionaryStack?.removeLastObject()
    }
    
    private func filterResponse(_ dictionnary: NSMutableDictionary?, key: String? = nil, count: Int = 0) -> NSMutableDictionary?{
        
        let filterdDic = NSMutableDictionary()
        
        if let keys = dictionnary?.allKeys as? [String] {
            for key in keys {
                if let value = dictionnary?.value(forKey: key) as? NSMutableDictionary {
                    
                    if value.count == 1 {
                        if let valueD = value.object(forKey: key) {
                            dictionnary?.setValue(valueD, forKey: key)
                        }
                    } else {
                        
                        let dic = removeKeyFromDic(value, keyToRemove: key)
                        let _ = filterResponse(dic, key: key, count: count+1)
                        filterdDic.setValue(dic, forKey: key)
                    }
                    
                    
                } else if let values = dictionnary?.value(forKey: key) as? [NSMutableDictionary] {
                    
                    for value in values {
                        if value.count == 1 {
                            if let valueD = value.object(forKey: key) {
                                dictionnary?.setValue(valueD, forKey: key)
                            }
                        } else {
                            
                            let dic = removeKeyFromDic(value, keyToRemove: key)
                            let _ = filterResponse(dic, key: key, count: count+1)
                            filterdDic.setValue(dic, forKey: key)
                        }
                    }
                }
            }
        }
        
        if count == 0 {
            return filterdDic
        } else {
            return nil
        }
    }
    
    private func removeKeyFromDic(_ dictionnary: NSMutableDictionary, keyToRemove: String) -> NSMutableDictionary {
        
        if let value = dictionnary.value(forKey: keyToRemove) as? String{
            if value.isEmpty == true{
                dictionnary.removeObject(forKey: keyToRemove)
            }
        }
    
        return dictionnary
    }
}


XMLParser/AppDelegate.swift
//
//  AppDelegate.swift
//  XMLParser
//
//  Created by Amitabha Saha on 25/06/18.
//  Copyright © 2018 ASTech. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }

    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and invalidate graphics rendering callbacks. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the active state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



XMLParser/ASXMLCreator.swift
//
//  ASXMLCreator.swift
//  XMLParser
//
//  Created by Amitabha Saha on 27/06/18.
//  Copyright © 2018 ASTech. All rights reserved.
//

import Foundation

class ASXMLCreator {
    
    typealias JSON = [String: Any]
    
    var input: JSON?
    var output: String?
    init(input: JSON) {
        
        self.input = input
        let xmlSt = createXML(self.input)
        output = "<?xml version=\"1.0\" encoding=\"UTF-8\" ?>" + "<root>" + xmlSt + "</root>"
    }
    
    
    private func createXML(_ dictionnary: JSON?, key: String? = nil) -> String {
        
        var xml = ""
        dictionnary?.keys.forEach({ (dickey) in
            if let value = dictionnary?[dickey] as? JSON {
                
                let xmlSt = createXML(value, key: dickey)
                xml = xml + "<\(dickey)>" + xmlSt + "</\(dickey)>"
                
            } else if let values = dictionnary?[dickey] as? [JSON] {
                
                for value in values {
                    let xmlSt = createXML(value, key: dickey)
                    xml = xml + "<\(dickey)>" + xmlSt + "</\(dickey)>"
                }
                
            } else if let value = dictionnary?[dickey] as? String {
                
                xml += "<\(dickey)>\(value)</\(dickey)>"
            }
        })
        
        return xml
    }
}


XMLParser/ViewController.swift
//
//  ViewController.swift
//  XMLParser
//
//  Created by Amitabha Saha on 25/06/18.
//  Copyright © 2018 ASTech. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
    }
    
    @IBAction func parseAction(_ sender: Any) {
        
        let education = [["course":"graduation",
                          "institutation":"IITD",
                          "from":"2007",
                          "to":"2011"],
                         ["course":"10+2",
                          "institutation":"DPS",
                          "from":"2005",
                          "to":"2007"],
                         ["course":"10",
                          "institutation":"DPS",
                          "from":"2000",
                          "to":"2005"]]
        let dic : [String : Any] = ["fisrtname":"John",
                                    "lastname":"Doe",
                                    "email":"john@doe.com",
                                    "age":"36",
                                    "gender":"male",
                                    "education":education]
        let xml = ASXMLCreator.init(input: dic)
        print("XML: \(xml.output ?? "")")
        
        if let path = Bundle.main.path(forResource: "Foods", ofType: "xml"){
            let url = URL(fileURLWithPath: path)
            let xmlString = try! String(contentsOf: url, encoding: String.Encoding.utf8)

            if let dictionary = ASXMLParser().dictionaryForXMLString(string: xmlString) {
                print(dictionary)
            }

        }
    }
    

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }


}

